### PR TITLE
Add SPA fallback, configurable extension, and refactor handler

### PR DESCRIPTION
**Pull request by akshat-kumar-singhal**
## Summary
- **SPA mode** (`SPA_MODE=true`): serves `index.html` with 200 for extensionless routes that don't match a file, enabling client-side routing without needing a `404.html` copy of `index.html`
- **Configurable default extension** (`DEFAULT_EXTENSION`): auto-resolve extensionless paths to any file type (defaults to `.html`)
- **Extract middleware** into testable `staticFileHandler` struct backed by GoFr's `file.FileSystem` interface
- Replace per-request `regexp.MatchString` with `filepath.Ext`
- Use `strconv.ParseBool` for `SPA_MODE` config
- Fix Dockerfile to build all Go files, not just `main.go`
- Update GoFr to v1.56.0, replace deprecated `file.New` with `file.NewLocalFileSystem`
- Replace `go main()` + `time.Sleep` in tests with `httptest.NewServer` for proper lifecycle
- Streamline README with config table and concise usage

## Bug fixes
- **404 response headers**: the previous handler called `w.WriteHeader(404)` before `http.ServeFile`, flushing headers early and preventing `ServeFile` from setting `Content-Type`, `Content-Length`, and caching headers. The new `statusOverrideWriter` lets `ServeFile` set headers normally and only overrides the status code.
- **Broader `Stat` error handling**: `ServeHTTP` now treats any `Stat` error (not just `os.IsNotExist`) as "not serveable" and routes to the 404 / SPA fallback, instead of falling through to `http.ServeFile` on a broken path.

## Test plan
- [x] `go test ./...` — tests pass across 2 packages
- [x] `go vet ./...` — clean
- [x] Docker: `SPA_MODE=true`, `/dashboard` → 200 with `index.html`
- [x] Docker: `DEFAULT_EXTENSION=.json`, `/data` → 200 with `data.json`
- [x] Docker: `/missing.css` → 404
- [x] Local: all above verified with GoFr v1.56.0
- [x] No zombie server processes after test runs

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by aryanmehrotra — COMMENTED** [review]
## Thorough Review — PR #21

**Build**: compiles clean, `go vet` clean, all 19 tests pass.

---

### Breaking-change risk for existing services

The frontends (`frontend/app/Dockerfile`, `frontend/website/Dockerfile`) currently pin `zopdev/static-server:v0.0.8`. The README bumps the example to `v0.0.9`. **As long as consumers stay on v0.0.8 they are unaffected.** However, when they upgrade to v0.0.9, behavior should be identical because:

- `SPA_MODE` defaults to `false` — existing behavior preserved.
- `DEFAULT_EXTENSION` defaults to `.html` — same as the old hardcoded `.html`.
- `CONFIG_FILE_PATH` hydration is unchanged.
- `/.well-known/` passthrough is unchanged.

**No existing service should break on upgrade** as long as the new env vars are left unset. The `.env` template adds `SPA_MODE=` and `DEFAULT_EXTENSION=` with empty values, and `strconv.ParseBool("")` returns `false`, which is correct.

---

### Issues

#### 1. `Stat` error handling is too narrow (Bug)
**`handler.go:29`** — `os.IsNotExist(err)` only catches "file not found". If `Stat` returns a *different* error (permission denied, I/O error, symlink loop), the code falls through to `http.ServeFile(w, r, filePath)` on line 41, which will also fail — but with an uncontrolled 5xx or misleading response. The old code had the same bug with `os.Stat`, but since you're refactoring anyway:

```go
if _, err := h.fs.Stat(filePath); err != nil {
```

would be safer. Any stat failure means the file isn't serveable.

#### 2. `resolveFilePath` leaks `hasExtension` semantic for directory paths
**`handler.go:47`** — `filepath.Ext("/blog/")` returns `""`, so `hasExtension` is false. But then the directory branch on line 54 resolves it to `blog/index.html`. If that `index.html` doesn't exist, we fall back to SPA mode and serve the app's root `index.html` — which is probably wrong for a directory URL. This is a minor edge case but worth a test.

#### 3. Missing test: SPA mode with `DEFAULT_EXTENSION` interaction
When `SPA_MODE=true` and `DEFAULT_EXTENSION=.html`:
- `/docs` → resolves to `docs.html` (exists) → serves `docs.html` ✓
- `/dashboard` → no `dashboard.html` → SPA fallback → serves `index.html` ✓

But what if `DEFAULT_EXTENSION=.json`?
- `/docs` → tries `docs.json` (doesn't exist) → tries `docs/` (not a dir) → falls through to `ServeHTTP` → stat fails → SPA fallback → serves `index.html`
- This means `/docs` would serve `index.html` instead of `docs.html` when extension is `.json` and SPA is on. This might surprise users. Worth documenting or testing explicitly.

#### 4. `statusOverrideWriter` swallows the original status silently
**`handler.go:67`** — The unnamed `int` parameter discards whatever status `http.ServeFile` sends. This works for the 404 case, but if `404.html` itself doesn't exist, `http.ServeFile` would internally want to send its own 404 + error body, and the override writer would still write 404 (correct status) but the response body might be `http.ServeFile`'s built-in "404 page not found" text instead of a styled page. Not a bug per se, but worth noting as expected behavior.

#### 5. `handler.fs` initialized in `OnStart`, not at construction
**`main.go:32`** — `handler.fs` is set inside `OnStart`. I verified GoFr's source — `OnStart` hooks run **before** the HTTP server starts accepting requests (`App.Run()` calls `handleStartupHooks()` synchronously before `startAllServers()`). So this is safe. But a code comment or moving `fs` into the constructor would make this less fragile for future readers.

#### 6. `handler.next` set inside `UseMiddleware` closure — timing concern
**`main.go:42`** — `handler.next` is assigned when GoFr calls the middleware factory. This happens during server setup, not per-request, so it's safe. But `next` is set by a side-effect of `UseMiddleware` rather than explicitly — if GoFr ever changes its middleware lifecycle, this would silently break. Consider passing `next` as a parameter to a constructor function instead.

---

### Minor / Non-blocking

1. **`go.mod` adds new indirect deps** (`graphql-go/graphql`, `vektah/gqlparser`, `agnivade/levenshtein`, `go.uber.org/atomic`) — these come from GoFr v1.56.0 pulling in GraphQL support. No concern, but the binary size will grow slightly.

2. **`configs/.env` missing trailing newline** — The `\ No newline at end of file` warning is still present. Not a functional issue, but some tools behave oddly without it.

3. **`TestServer` in `main_test.go` and `TestServeHTTP` in `handler_test.go` overlap** — Both test similar scenarios (existing file, missing file, extension resolution). Consider whether `TestServer` (which spins up a real `httptest.Server`) adds value beyond `TestServeHTTP` (which uses `httptest.NewRecorder`). The `TestServer` is slightly more integration-y since it goes through real HTTP, but the coverage is largely redundant.

4. **`setupTestDir` uses map iteration** — File creation order is non-deterministic. Doesn't matter for correctness, but switching to a slice of structs would make the test more readable and deterministic (same pattern the old code used).

5. **Dockerfile uses `golang:1.26`** — Just confirming this is the intended Go version for the build. If the project CI also runs on this version, no issue.

6. **README references `v0.0.9`** — This version tag doesn't exist yet. Make sure to tag after merge, or consumers referencing this README will pull a non-existent image.

---

### What's good

- **Replaced per-request `regexp.MatchString` with `filepath.Ext`** — eliminates regex compilation overhead on every request.
- **Replaced `os.Stat` with `file.FileSystem.Stat`** — testable without touching the real filesystem.
- **Killed `go main()` + `time.Sleep(3s)`** — the old test was non-deterministic and could leave zombie processes. `httptest.NewServer` is the right pattern.
- **`statusOverrideWriter`** is a strict improvement over the old `w.WriteHeader(404)` followed by `http.ServeFile`. The old approach sent headers prematurely, preventing `http.ServeFile` from setting `Content-Type` and `Content-Length`. The new wrapper lets `ServeFile` set all headers normally and only overrides the status code.
- **SPA mode and configurable extension** are clean, opt-in additions with safe defaults.
- **README is much more concise** — the config table is a clear improvement.

---

### Verdict

**Approve with nits.** The fix in #1 (`os.IsNotExist` → `err != nil`) is the only one I'd consider blocking — everything else is minor. No existing services will break on upgrade since new features are opt-in with backward-compatible defaults.

File: handler.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+func (h *staticFileHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if strings.Contains(r.URL.Path, "/.well-known/") {
+		h.next.ServeHTTP(w, r)
+
+		return
+	}
+
+	filePath, hasExtension := h.resolveFilePath(r.URL.Path)
+
+	if _, err := h.fs.Stat(filePath); os.IsNotExist(err) {
```
Comment (aryanmehrotra) [inline handler.go]: `os.IsNotExist(err)` only catches "file not found". If Stat fails for any other reason (permissions, I/O error, symlink loop), we skip the 404 branch and fall through to `http.ServeFile` on line 41 with a broken path — uncontrolled error response. Should just be `err != nil` here.

File: handler.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+func (h *staticFileHandler) resolveFilePath(urlPath string) (string, bool) {
+	filePath := filepath.Join(h.staticFilePath, urlPath)
+
+	hasExtension := filepath.Ext(filePath) != ""
+
+	if urlPath == rootPath {
+		filePath += indexHTML
+	} else if !hasExtension {
+		if _, err := h.fs.Stat(filePath + h.defaultExtension); err == nil {
```
Comment (aryanmehrotra) [inline handler.go:51]: When `DEFAULT_EXTENSION=.json` and `SPA_MODE=true`: a request to `/docs` tries `docs.json` (not found), then checks if `docs/` is a dir (no) — and falls into the SPA fallback, serving `index.html` instead of the existing `docs.html`. Might be surprising. Worth either documenting this or always trying `.html` as a secondary fallback.

File: handler.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	filePath, hasExtension := h.resolveFilePath(r.URL.Path)
+
+	if _, err := h.fs.Stat(filePath); os.IsNotExist(err) {
+		if h.spaMode && !hasExtension {
+			http.ServeFile(w, r, filepath.Join(h.staticFilePath, indexHTML))
+			return
+		}
+
+		http.ServeFile(&statusOverrideWriter{ResponseWriter: w, status: http.StatusNotFound}, r,
```
Comment (aryanmehrotra) [inline handler.go:34]: Nice fix on the old 404 handling — the previous code called `w.WriteHeader(404)` before `http.ServeFile`, which flushed headers early and prevented ServeFile from setting `Content-Type`, `Content-Length`, and caching headers. The `statusOverrideWriter` approach fixes that. Worth mentioning in the PR description as a bug fix, not just a refactor.

**Review by aryanmehrotra — COMMENTED** [review]
Posting focused line comments below. Please ignore the earlier wall-of-text review.

**Review by aryanmehrotra — APPROVED** [review]
